lib/runner.js

```diff
@@ -1027,11 +1004,34 @@ function extraGlobals() {
     });
 
     // 'errno' was renamed to process._errno in v0.9.11.
-
     if (nodeVersion < 0x00090b) {
       return ['errno'];
     }
   }
 
   return [];
 }
+
+/**
+ * Constants used by Runner
+ * @public
+ */
+var constants = Object.freeze({
+  RUNNER_EVENT_END: 'end',
+  RUNNER_EVENT_FAIL: 'fail',
+  RUNNER_EVENT_HOOK: 'hook',
+  RUNNER_EVENT_HOOK_END: 'hook end',
+  RUNNER_EVENT_PASS: 'pass',
+  RUNNER_EVENT_PENDING: 'pending',
+  RUNNER_EVENT_RETRY: 'retry',
+  RUNNER_EVENT_BEGIN: 'start',
+  RUNNER_EVENT_SUITE: 'suite',
+  RUNNER_EVENT_SUITE_END: 'suite end',
+  RUNNER_EVENT_TEST: 'test',
+  RUNNER_EVENT_TEST_END: 'test end',
+  RUNNER_EVENT_WAITING: 'waiting',
+  RUNNER_STATE_FAILED: 'failed',
+  RUNNER_STATE_PASSED: 'passed'
+});
+
+module.exports.constants = constants;
```

Comment:
Consider instead:

**"lib/events.js"**
```js
module.exports = {
  Runnable: {
    EVENT_RUNNABLE_ERROR: 'error'
  },
  Suite: {
    EVENT_AFTER_ALL_HOOK_QUEUED: 'afterAll',
    EVENT_AFTER_EACH_HOOK_QUEUED: 'afterEach',
    EVENT_BEFORE_ALL_HOOK_QUEUED: 'beforeAll',
    EVENT_BEFORE_EACH_HOOK_QUEUED: 'beforeEach',

    EVENT_POST_REQUIRE: 'post-require',
    EVENT_PRE_REQUIRE: 'pre-require',
    EVENT_REQUIRE: 'require',

    EVENT_ROOT_SUITE_RUN: 'run',

    EVENT_SUITE_QUEUED: 'suite',
    EVENT_TEST_QUEUED: 'test'
  },
  Runner: {
    EVENT_HOOK_BEGIN: 'hook',
    EVENT_HOOK_END: 'hook end',

    EVENT_ROOT_SUITE_BEGIN: 'start',
    EVENT_ROOT_SUITE_END: 'end',
    EVENT_ROOT_SUITE_WAITING: 'waiting',

    EVENT_SUITE_BEGIN: 'suite',
    EVENT_SUITE_END: 'suite end',

    EVENT_TEST_BEGIN: 'test',
    EVENT_TEST_END: 'test end',

    EVENT_TEST_PASS: 'pass',
    EVENT_TEST_FAIL: 'fail',
    EVENT_TEST_PENDING: 'pending',
    EVENT_TEST_RETRY: 'retry'
  }
};
```

Then in usage:
```js
var events = require('lib/events').Runner;
...
  runner.on(events.EVENT_TEST_BEGIN, ...);
```

The events for each should be JSDoc'ed (not shown above).
```js
/**
 * Test fail event.
 * Indicates test failed.
 *
 * @event module:Runner.event:fail
 * @property {Test} test - Instance that was executed
 * @property {Error} err - Cause of failure
 */
```

Reply:
I'm following Node.js' lead on how they export constants.  Each module contains a `constants` object which contains constants relevant to that module.

Reply:
[Such as](https://github.com/search?q=constants++repo%3Anodejs%2Fnode+path%3A%2Flib+language%3AJavaScript&type=Code&ref=advsearch&l=JavaScript&l=)?

Reply:
`fs.constants`, `os.constants`, `zlib.constants`, `http2.constants`, etc.

Reply:
- [ ] add docstrings for all events

Reply:
Given your examples, create `Runner.constants.events` mirroring how they handle `Error.codes`.

Reply:
Had an exact example but browser crashed.
But `os` module provides similar division of `constants`.

```js
var os = require('os')
os.constants.errno
os.constants.signals
os.constants.priority
```

Reply:
I see

Reply:
OK, so I can do that, but what I still don't want to do is stuff everything in one file.  that should be a reasonable compromise

Reply:
but we still need the prefixes for autocompletion.

Reply:
bah, actually, some modules do it `os`-style, and others do it `fs`-style, and I'm tired of working on this PR, so I'm going to leave it.  if we nest them it'll be more annoying for users to reference them anyway.